2025-05-99-how-to-implement-zero-copy-communication/index.md

```diff
@@ -0,0 +1,164 @@
+
+```rust
+let data_ptr = sensor_get_next_frame();
+```
+
+However, this is not zero-copy: the data resides in sensor-managed memory, and
+must be copied into shared memory before sending. For true zero-copy, we
+need the sensor to write directly into shared memory, provided by the
+zero-copy communication framework.
+
+* **1. Produce Data into Shared Memory**
+  First, we acquire an uninitialized sample from the publisher. This is a chunk
+  of memory from the publisher’s data segment, reserved for the user's payload.
+  Under the hood, an allocator manages this memory block:
+
+  ```rust
+  let mut uninitialized_sample =
+      publisher.loan_slice_uninit(MAX_SIZE_OF_SENSOR_DATA_FRAME)?;
+  ```
+
+  Next, the sensor writes directly into this memory:
+
+  ```rust
+  sensor_get_next_frame(uninitialized_sample.payload_mut());
+  ```
+
+  * **Deep Dive:**
+    When the publisher is created:
+    1. A shared memory object is created using `shm_open`.
+    2. It is resized via `ftruncate` to reserve enough space.
+    3. The segment is mapped into the process using `mmap`.<br><br>
+
+    In iceoryx2 the memory of the data segment is managed by a pool allocator.
+    It partitions memory into fixed-size regions, which avoids fragmentation
+    and allows predictable allocation times - essential for real-time and
+    safety-critical systems.
+
+    The call `loan` allocates a chunk of memory from this pool allocator and
+    returns it to the user.
+
+* **2. Send Pointer To Data**
+  Once the data is produced, the sample is marked as initialized and published:
+
+  ```rust
+  let sample = unsafe { uninitialized_sample.assume_init(); }
+  sample.send()?;
+  ```
+
+  * **Deep Dive:**
+    Only an offset to the memory location is sent to the subscribers, not the
+    actual data.
+    Each process has its own virtual memory layout, so we cannot directly
+    share pointers. Instead:
+    * The publisher calculates an offset from the base of the shared memory segment.
+    * The subscriber adds this offset to its local mapping of that segment.
+    * A valid pointer to the received data is reconstructed.<br><br>
+
+    To transfer the offset, an inter-process communication mechanisms like
+    message queues, pipes, unix domain sockets, or shared memory queues can be
+    used.
+
+    <table width=100% height=100% border=0><tr><td align=center>
+
+    <img src="offset.png" alt="offset" style="max-width: 80%"/>
+
+    </td></tr></table>
+
+
+    <br><br>
+
+* **3. Read Data**
+  Upon receiving the offset, the subscriber can reconstruct a pointer and
+  access the data directly:
+
+  ```rust
+  if let Some(sample) = subscriber.receive()? {
+    process_sensor_data(sample.payload());
+  }
+  ```
+
+  * **Deep Dive:**
+    Before a subscriber can consume the received data:
+    1. It must map the publisher’s shared memory using `shm_open` and `mmap`.
+    2. It listens for incoming offsets from the publisher.
```

Comment:
Not sure if `listen` is the best wording here, since we also have a `listener`. How about `It polls for incoming offsets` and as a side note `If polling is not desired, the event messaging pattern can be combined with publish-subscribe`?